keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -0,0 +1,258 @@
+
+Example credspec.json
+```
+{
+  "CmsPlugins": [
+    "ActiveDirectory"
+  ],
+  "DomainJoinConfig": {
+    "DnsName": "contoso.com",
+    "Guid": "244818ae-87ca-4fcd-92ec-e79e5252348a",
+    "DnsTreeName": "contoso.com",
+    "NetBiosName": "DEMO",
+    "Sid": "S-1-5-21-2126729477-2524075714-3094792973",
+    "MachineAccountName": "WebApplication1"
+  },
+  "ActiveDirectoryConfig": {
+    "GroupManagedServiceAccounts": [
+      {
+        "Name": "WebApplication1",
+        "Scope": "DEMO"
+      },
+      {
+        "Name": "WebApplication1",
+        "Scope": "contoso.com"
+      }
+    ]
+  }
+}
+```
+
+  - Create a configmap in the namespace with the contents of the JSON - `kubectl create configmap webserver-credspec --from-file=gmsa-cred-spec=/path/to/credspec.json`
+
+3. Secure the configmap with the GMSA credspec so that only desired users/service accounts within a namespace can view, modify or use the GMSA configmap:
+  - Create a role to view, modify or use the GMSA configmap:
+
+```
+kind: Role
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  namespace: default
+  name: webserver-gmsa-user
+rules:
+- apiGroups: [""]
+  resources: ["configmaps"]
+  resourceNames: ["webserver-credspec"]
+  verbs: ["get", "patch", "delete", "update", "use"]
+```
+
+  - Create a rolebinding and assign desired users (for example, jane below) or service accounts to the above role:
+
+```
+kind: RoleBinding
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: use-webserver-gmsa
+  namespace: default
+subjects:
+- kind: User
+  name: jane
+  apiGroup: rbac.authorization.k8s.io
+roleRef:
+  kind: Role
+  name: webserver-gmsa-user
+  apiGroup: rbac.authorization.k8s.io
+```
+  
+4. Application admin deploys the application pods with `spec.securityContext.windows.credentialSpecConfig` set to the name of any authorized GMSA configmap (e.g. `webserver-credspec`) as part of the PodSpec
+5. During pod creation, a new Admission Controller, GMSAAuthorizer, will ensure the `user` (and `spec.serviceAccountName` if specified) is authorized for the `use` verb on the GMSA configmap. If the authorization check fails due to absence of RBAC roles, the pod creation fails.
+6. The Windows CRI implementation accesses the configmap with the GMSA credspec and copies the contents of entry with key `gmsa-cred-spec` to the [OCI windows.CredentialSpec](https://github.com/opencontainers/runtime-spec/blob/master/config-windows.md#credential-spec) field
+7. The Windows OCI implementation validates the credspec and fails to start the container if its invalid or access to the GMSA from the node is denied
+8. The container starts with the Windows Group Managed Service Account as expected, and can authenticate to the database.
+9. During any subsequent pod update, GMSAAuthorizer blocks any changes to a pod's `spec.securityContext.windows.credentialSpecConfig`
```

Comment:
Why is this needed? Does it also preclude updating the gMSA used by say a Deployment?

Reply:
I wanted to block updates for a podspec around GMSA but seems like this is already handled at https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L3563 . So I will remove any extra work around blocking pod spec updates. It should not preclude updating the gMSA used by a Deployment/etc. since we definitely do want to enable  updates for the gMSA config there.